File: programs/benchmarks-alioth/python/binarytrees.py
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org/
#
# contributed by Antoine Pitrou
# modified by Dominique Wahli
# modified by Heinrich Acker
# adapted for p2w
#
# Single-threaded version optimized for p2w


def make_tree(depth: int):
    if depth == 0:
        return (None, None)
    depth = depth - 1
    return (make_tree(depth), make_tree(depth))


def check_tree(node) -> int:
    left = node[0]
    right = node[1]
    if left is None:
        return 1
    return 1 + check_tree(left) + check_tree(right)


def main(n: int) -> None:
    min_depth: int = 4
    max_depth: int = 0
    if min_depth + 2 > n:
        max_depth = min_depth + 2
    else:
        max_depth = n

    stretch_depth: int = max_depth + 1

    stretch_tree = make_tree(stretch_depth)
    stretch_check: int = check_tree(stretch_tree)
    print("stretch tree of depth", stretch_depth, "check:", stretch_check)

    long_lived_tree = make_tree(max_depth)

    iterations: int = 1
    i: int = 0
    while i < max_depth:
        iterations = iterations * 2
        i = i + 1

    depth: int = min_depth
    while depth < stretch_depth:
        check: int = 0
        i = 1
        while i <= iterations:
            check = check + check_tree(make_tree(depth))
            i = i + 1

        print(iterations, "trees of depth", depth, "check:", check)
        iterations = iterations // 4
        depth = depth + 2

    print("long lived tree of depth", max_depth, "check:", check_tree(long_lived_tree))


# Default argument for testing
main(10)

File: programs/benchmarks-alioth/zig/binarytrees.zig
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// Zig port (Zig 0.15+)

const std = @import("std");
const posix = std.posix;

fn print(comptime fmt: []const u8, args: anytype) void {
    var buf: [256]u8 = undefined;
    const msg = std.fmt.bufPrint(&buf, fmt, args) catch unreachable;
    _ = posix.write(posix.STDOUT_FILENO, msg) catch {};
}

const TreeNode = struct {
    left: ?*TreeNode,
    right: ?*TreeNode,
};

fn bottomUpTree(allocator: std.mem.Allocator, depth: u32) !*TreeNode {
    const node = try allocator.create(TreeNode);
    if (depth > 0) {
        node.left = try bottomUpTree(allocator, depth - 1);
        node.right = try bottomUpTree(allocator, depth - 1);
    } else {
        node.left = null;
        node.right = null;
    }
    return node;
}

fn itemCheck(tree: *const TreeNode) i64 {
    if (tree.left) |left| {
        return 1 + itemCheck(left) + itemCheck(tree.right.?);
    } else {
        return 1;
    }
}

fn deleteTree(allocator: std.mem.Allocator, tree: *TreeNode) void {
    if (tree.left) |left| {
        deleteTree(allocator, left);
        deleteTree(allocator, tree.right.?);
    }
    allocator.destroy(tree);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var args = std.process.args();
    _ = args.skip();

    const n_str = args.next() orelse "10";
    const n = std.fmt.parseInt(u32, n_str, 10) catch 10;

    const min_depth: u32 = 4;
    const max_depth = if (min_depth + 2 > n) min_depth + 2 else n;
    const stretch_depth = max_depth + 1;

    // Stretch tree
    {
        const stretch_tree = try bottomUpTree(allocator, stretch_depth);
        print("stretch tree of depth {d}\t check: {d}\n", .{ stretch_depth, itemCheck(stretch_tree) });
        deleteTree(allocator, stretch_tree);
    }

    // Long-lived tree
    const long_lived_tree = try bottomUpTree(allocator, max_depth);

    var depth = min_depth;
    while (depth <= max_depth) : (depth += 2) {
        const iterations = @as(i64, 1) << @intCast(max_depth - depth + min_depth);
        var check: i64 = 0;

        var i: i64 = 1;
        while (i <= iterations) : (i += 1) {
            const temp_tree = try bottomUpTree(allocator, depth);
            check += itemCheck(temp_tree);
            deleteTree(allocator, temp_tree);
        }

        print("{d}\t trees of depth {d}\t check: {d}\n", .{ iterations, depth, check });
    }

    print("long lived tree of depth {d}\t check: {d}\n", .{ max_depth, itemCheck(long_lived_tree) });
    deleteTree(allocator, long_lived_tree);
}
